VPA: (InPlaceOrRecreate) Allow updater to actuate InPlaceOrRecreate updates

maxcao13 opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Introduces large changes in the updater component to allow `InPlaceOrRecreate` mode. If the feature gate is enabled and the VPA update mode is `InPlaceOrRecreate`, the updater will attempt an in place update by first checking a number of preconditions before actuation (e.g., whether we are already in-place resizing,
whether an in-place update may potentially violate disruption(previously eviction) tolerance, etc.). After the preconditions are validated, we send an update signal to the `InPlacePodVerticalScaling` API with the recommendation, which may or may not fail. Failures are handled in subsequent updater loops.

We need this in order to actually actuate in place updates.

#### Which issue(s) this PR fixes:
Part of [AEP-4016](https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler/enhancements/4016-in-place-updates-support) (InPlaceVerticalScaling/InPlaceOrRecreate)
This PR is part of the larger feature PR in https://github.com/kubernetes/autoscaler/pull/7673

Depends on:
- https://github.com/kubernetes/autoscaler/pull/7934
- https://github.com/kubernetes/autoscaler/pull/7961
### Note that this PR is merging into the [in-place-updates](https://github.com/kubernetes/autoscaler/tree/in-place-updates) feature branch, which will be merged when this feature is all reviewed and ready.

#### Special notes for your reviewer:
As for implementation details, `patchCalculators` have been re-used from the admission-controllers code for the updater in order to calculate recommendations for the updater to actuate. InPlace logic has been mostly stuffed in the eviction package for now because of similarities and ease (user-initated API calls eviction vs. in-place; both cause disruption), maybe we also need a "disruption limiter". It may or may not be useful to refactor this later. 

There are some TODOs littered throughout the code. I've limited as many TODOS as possible, but I'm using the rest of the TODOs as review markers to draw attention to parts of the discussion that need community discussion. Otherwise, they are items that should be fixed after 1.33 1287 KEP changes are released, or other future work.

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
[AEP-4016]: https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler/enhancements/4016-in-place-updates-support
```

maxcao13 commented:
/hold

upodroid commented:
/test pull-kubernetes-e2e-autoscaling-vpa-full

adrianmoisey commented on vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go:
```diff
@@ -392,3 +431,149 @@ func setUpInformer(kubeClient kube_client.Interface, kind controllerKind) (cache
+	}
+	cr, present := e.podToReplicaCreatorMap[GetPodID(pod)]
+	if present {
+		// If our QoS class is guaranteed, we can't change the resources without a restart
```
Why not?

Taking a look at https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources#static-cpu--memory-policy, it seems as though it can't be resized if QoS is guaranteed and the node has a static CPU or memory policy configured

Is that the reason why?

maxcao13 replied:
> If a pod is a Guaranteed, then it basically means all requests/limits are set and are equal.
If the VPA is set to control both requests and limits, then the Pod will continue to stay in Guaranteed.

In `Recreate` mode, not necessarily since we are evicting, the qos class doesn't matter.
In `InPlaceOrRecreate`, I think any attempt to change requests and limits where the requests != limits will fail and force VPA to evict.

> If the VPA is requests-only, then the Pod can't be resized down (as that changes the QoS) or up (we're at the max anyway, since requests can't be more than limits).

In place resizes here won't work either I believe, and we would fallback to eviction.

> So we can resize a Guaranteed pod, as long as it's to make it larger?

As long as every container in the pod has both the limits and requests are equal after an in-place update finished (i.e. qos class doesn't change from guaranteed), then the resize will fail.

I guess we could just remove this check if we think that would happen more than often? But the chance seems very, very small just going off of recommendations, and I'm not aware of a VPA recommendation feature which preserves `Guranteed` qos class.

raywainman replied:
Today in VPA, in the case where the pod uses a `Guaranteed` QOS, wouldn't it still preserve this during an update of resources because requests and limits would just both be updated to the same value, thereby preserving the pod's QOS?

My understanding is the only time requests and limits diverge is when the pod is burstable and VPA tries to maintain the ratio between request and limits as it updates the values. In the Guaranteed QOS case, the ratio is 1 and so that should just be preserved.

So my (quick, unchecked) intuition here is that we can remove this check?

adrianmoisey commented on vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go:
```diff
@@ -392,3 +431,149 @@ func setUpInformer(kubeClient kube_client.Interface, kind controllerKind) (cache
+		}
+
+		singleGroupStats, present := e.creatorToSingleGroupStatsMap[cr]
+		// If we're pending, we can't in-place resize
+		// TODO(jkyros): are we sure we can't? Should I just set this to "if running"?
```
Good question, I imagine that a resize during pending is "less risky" as there's no running software/cgroup.

maxcao13 replied:
Right, I'm not exactly sure what happens if you try this though. I will report back to see what happens, and act accordingly.

In theory we would want behaviour in a situation e.g. 
3 replica deployment
2 of them are running, but 1 is pending, and the 2 replicas are generating good recommendations that the pending pod would like to use while pending.

I'm assuming that's why this would be useful?

adrianmoisey replied:
> The thing is that the VPA updater seems to ignore pending pods for updates anyways, so I guess there's no point in this check at all? ([The podLister does not list pending pods](https://github.com/kubernetes/autoscaler/blob/52cd68a4987f875eb811afbe07b66d9d69f51a14/vertical-pod-autoscaler/pkg/updater/logic/updater.go#L176) at least for my kind cluster). Am I missing something? I thought the updater recognizes pending pods, since that's a condition in `CanEvict`...

Taking a look at this: https://github.com/kubernetes/autoscaler/blob/aa1d413ea3bf319b56c7b2e65ade1a028e149439/vertical-pod-autoscaler/pkg/updater/logic/updater.go#L299-L310

I assume that `"spec.nodeName!=" + ""` may exclude some Pending pods?

I wonder if this is something to change in the future?

raywainman replied:
Yes, my understanding too is that `"spec.nodeName!=" + ""` will exclude pending pods.

maxcao13 replied:
I think I should remove this check as well then.

omerap12 commented on vertical-pod-autoscaler/pkg/utils/metrics/updater/updater.go:
```diff
@@ -76,13 +76,45 @@ var (
+			Help:      "Number of VPA objects with at least one in-place updated Pod.",
+		}, []string{"vpa_size_log2"},
+	)
+
```
Do we want to add another metric for failed in-place autoscaling attempts?

omerap12 replied:
It can be addressed in a follow-up PR, but just a thought - this metric might be useful for the end user.

maxcao13 replied:
That's a good idea. There are many ways an in-place update could fail that aren't related. E.g. `qosClass` change, status infeasible, invaid rbac for patching, etc. Are we okay with just adding a singular, generic, catch-all metric for now?

omerap12 replied:
Yeah, I think so. Just add a TODO comment, and once this PR is merged, I'll open an issue to address it :)

raywainman commented on vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go:
```diff
@@ -392,3 +433,148 @@ func setUpInformer(kubeClient kube_client.Interface, kind controllerKind) (cache
+		}
+
+		for _, container := range pod.Spec.Containers {
+			// If some of these are populated, we know it at least understands resizing
```
@maxcao13 have you seen this behavior? Does this field only get populated AFTER the first resize or should it be there from the start of a pod's lifetime?

maxcao13 replied:
From what I've seen is that this field is always populated when a pod is created with the feature gate on.

maxcao13 replied:
Seems like from here: https://github.com/kubernetes/kubernetes/blob/83bb5d570580a3f477737fec5c24ba8fc3554264/pkg/api/pod/util.go#L638

If the feature gate is not enabled, the field on created pods gets dropped. So it's probably a good indication that this is the case.

raywainman commented on vertical-pod-autoscaler/pkg/updater/logic/updater.go:
```diff
@@ -201,29 +222,60 @@ func (u *updater) RunOnce(ctx context.Context) {
-		evictionLimiter := u.evictionFactory.NewPodsEvictionRestriction(livePods, vpa)
-		podsForUpdate := u.getPodsUpdateOrder(filterNonEvictablePods(livePods, evictionLimiter), vpa)
+		evictionLimiter := u.evictionFactory.NewPodsEvictionRestriction(livePods, vpa, u.patchCalculators)
+		podsForUpdate := u.getPodsUpdateOrder(filterNonUpdatablePods(livePods, evictionLimiter), vpa)
```
Would it make things cleaner if we split out the pods that can be in-place updated?

```
podsForInPlaceUpdates := u.getPodsUpdateOrder(filterNonUpdatablePods(livePods, evictionLimiter), vpa)
podsForEviction := u.getPodsUpdateOrder(filterNonEvictablePods(livePods, evictionLimiter), vpa)
```

Then below you can treat them as two different cases with two different for-loops.

maxcao13 replied:
I've refactored it a bit, lmk if it makes sense.

raywainman reviewed (commented):
First pass through. Need to do another pass but thought I'd shoot off some questions to start.

Lots of great work here, thanks a lot @maxcao13!

maxcao13 commented on vertical-pod-autoscaler/deploy/vpa-rbac.yaml:
```diff
@@ -121,6 +121,32 @@ rules:
+      - ""
+    resources:
+      - pods/resize
+      - pods
```
`pods/resize` is sufficient for _resizing_, but I have a change in here that allows the updater to directly patch annotations on the pod: https://github.com/kubernetes/autoscaler/pull/7962/files#diff-85c8216ee51fe203ea35cff7dd231b0810bae6438fcfee5b9ee20cd2b7593671R525 and it's purpose is to signal to a user whether a pod/container was attempted for in-place vertical scaling. The admission-controller patches its own annotations directly to the pods as well ([vpaObservedContainers](https://github.com/kubernetes/autoscaler/blob/27d00f486c09d47ce9ae0aad680d3803519b61d7/vertical-pod-autoscaler/pkg/admission-controller/resource/pod/patch/observed_containers.go#L30)), albeit that this annotation actually serves a functioning purpose in the code. The annotation that _I_ am patching doesn't do anything functionally useful and requires this additional rbac rule.

It's purely for cosmetic reasons as noted in this comment: https://github.com/kubernetes/autoscaler/pull/7673#discussion_r1910002910, but the `vpaObservedContainers` annotation is actually used in GetUpdatePriority. If people don't think this is useful to users, then we could get rid of it entirely.

maxcao13 commented on vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go:
```diff
@@ -99,21 +112,42 @@ type podReplicaCreator struct {
+				if (actuallyAlive - 1) > shouldBeAlive { // -1 because this pod is the one being in-place updated
+					if pod.Status.Resize == apiv1.PodResizeStatusInfeasible || pod.Status.Resize == apiv1.PodResizeStatusDeferred {
+						klog.InfoS("Attempted in-place resize was impossible, should now evict", "pod", klog.KObj(pod), "resizePolicy", pod.Status.Resize)
+						return true
```
We _could_ check again here, but what happens in practice (I think) is that since we do `AttemptInPlaceUpdate` _first_ if the updateMode is `InPlaceOrRecreate`, the checks in there will already do the `Deferred` + `InProgress` cool-off period checking and return `fallbackToEviction` based on the checks or not. Then `CanEvict` check is not aware of these cool-off periods, and maybe it shouldn't because `CanEvict` can be used outside of the context of `InPlaceOrRecreate` while `AttemptInPlaceUpdate/InPlaceOrRecreate` specifically has to work within the cool-off rules, if that makes sense.

But anyways, that's what happens as of now. WDYT?

adrianmoisey commented on vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go:
```diff
@@ -392,3 +433,138 @@ func setUpInformer(kubeClient kube_client.Interface, kind controllerKind) (cache
+		if err != nil {
+			return err
+		}
+		klog.V(4).InfoS("In-place patched pod /resize subresource using patches ", "pod", klog.KObj(res), "patches", string(patch))
```
Major nit:
```suggestion
		klog.V(4).InfoS("In-place patched pod /resize subresource using patches", "pod", klog.KObj(res), "patches", string(patch))
```

maxcao13 commented on vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go:
Sure, I can do that. Thanks!

maxcao13 replied:
Please take a look at https://github.com/kubernetes/autoscaler/commit/5a3f6749ccbdeacae3f78f3dd9c6f0a3d50c17c2

maxcao13 commented on vertical-pod-autoscaler/pkg/updater/logic/updater.go:
```diff
@@ -326,3 +397,61 @@ func newEventRecorder(kubeClient kube_client.Interface) record.EventRecorder {
+	}
+
+	// TODO(jkyros): need our own rate limiter or can we freeload off the eviction one?
+	err = u.evictionRateLimiter.Wait(ctx)
```
True, makes sense to me.

maxcao13 replied:
Solved in https://github.com/kubernetes/autoscaler/commit/5a3f6749ccbdeacae3f78f3dd9c6f0a3d50c17c2, but I'm wondering if we should introduce more command line flags specifically for this rate limiter, or if we want to rename the existing eviction limiter rate flags and slowly deprecate the old flag names.

Or we just use the existing eviction limiter rate flags and ignore the fact that the name is now not entirely correct :-)

maxcao13 commented on vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go:
```diff
@@ -99,21 +112,42 @@ type podReplicaCreator struct {
+				"evicted", singleGroupStats.evicted,
+				"updating", singleGroupStats.inPlaceUpdating)
+			if IsInPlaceUpdating(pod) {
+				if (actuallyAlive - 1) > shouldBeAlive { // -1 because this pod is the one being in-place updated
```
Hm... thanks for bringing that to my attention. So I think you're right. If tolerance turns out to be 0, there is a path where in-placing pods can never get evicted because this path will return false.

I see why this is why this branch of code is useful:
```
		if singleGroupStats.running == singleGroupStats.configured &&
				singleGroupStats.evictionTolerance == 0 &&
				singleGroupStats.evicted == 0 {
				singleGroupStats.evicted == 0 &&
				singleGroupStats.inPlaceUpdating == 0 {
				return true
			}
```
since it accounts for the fact that tolerance could be 0.

I'll add another check like this to solve this, and refactor it so it's not ugly.

maxcao13 replied:
Hold on, I still have a commit that will try and address this. The current implementation is wrong and blocks eviction forever. Unit tests are helpful here :-)

maxcao13 commented on vertical-pod-autoscaler/pkg/updater/logic/updater.go:
```diff
@@ -326,3 +397,61 @@ func newEventRecorder(kubeClient kube_client.Interface) record.EventRecorder {
 	return eventBroadcaster.NewRecorder(vpascheme, apiv1.EventSource{Component: "vpa-updater"})
 }
+
+func (u *updater) AttemptInPlaceUpdate(ctx context.Context, vpa *vpa_types.VerticalPodAutoscaler, pod *apiv1.Pod, evictionLimiter eviction.PodsEvictionRestriction) (fallBackToEviction bool, err error) {
```
I just stuck all the old "attempt" logic into `CanEvict` and makes it return an enum instead of a bool. Hope that makes sense too and brings it more parallel to the `PodsEvictionRestriction` impl.

maxcao13 commented:
I refactored a lot of the in-place stuff out of the eviction code and separated them in https://github.com/kubernetes/autoscaler/commit/5a3f6749ccbdeacae3f78f3dd9c6f0a3d50c17c2.

I need to write more unit tests for the inplaceRestriction stuff but wanted to push this for now for review.

maxcao13 commented:
Sorry if it looks like the latest change was big, a lot of the changes were moving things around, renaming, lowercasing some fields, and adding/fixing unit tests.

Other than that, I've done a bunch of refactoring of tests and the tolerance code, and changed it so that `CanEvict` take into accoun about the in-place fallback to evict conditions.

I've added some unit tests to confirm the functionality.

Reviewers, please take a look when you can.

raywainman commented on vertical-pod-autoscaler/pkg/updater/restriction/pods_inplace_restriction.go:
```diff
@@ -0,0 +1,176 @@
+// CanInPlaceUpdate checks if pod can be safely updated
+func (ip *PodsInPlaceRestrictionImpl) CanInPlaceUpdate(pod *apiv1.Pod) utils.InPlaceDecision {
+	if !features.Enabled(features.InPlaceOrRecreate) {
+		return utils.InPlaceDeferred
```
should this be evict instead?

raywainman commented on vertical-pod-autoscaler/pkg/updater/inplace/inplace_recommendation_provider.go:
```diff
@@ -0,0 +1,84 @@
+
+// GetContainersResourcesForPod returns recommended request for a given pod.
+// The returned slice corresponds 1-1 to containers in the Pod.
+func (p *inPlaceRecommendationProvider) GetContainersResourcesForPod(pod *core.Pod, vpa *vpa_types.VerticalPodAutoscaler) ([]vpa_api_util.ContainerResources, vpa_api_util.ContainerToAnnotationsMap, error) {
```
I don't feel super great about most of this code being copied from the other provider... All I can spot here is that we removed the annotations generated by the other implementation, is there anything else?

If not, let's just use the canonical provider instead?

(I totally can see how this will be useful in the future but maybe we keep it simple for now?)

raywainman reviewed (commented):
Looking really good. Generally just some nits. Thanks Max!

maxcao13 commented on vertical-pod-autoscaler/pkg/updater/main.go:
```diff
@@ -188,6 +190,11 @@ func run(healthCheck *metrics.HealthCheck, commonFlag *common.CommonFlags) {
 
+	inPlaceRecommendationProvider := inplace.NewInPlaceRecommendationProvider(limitRangeCalculator, vpa_api_util.NewCappingRecommendationProcessor(limitRangeCalculator))
+
+	// TODO(maxcao13): figure out if we need to use NewInPlaceUpdatedCalculator; does it help the user to know if their pod was updated in-place as an annotation?
```
Forgot to remove this comment, thanks! :-)

maxcao13 commented on vertical-pod-autoscaler/pkg/updater/logic/updater.go:
```diff
@@ -90,19 +96,33 @@ func NewUpdater(
 	evictionRateLimiter := getRateLimiter(evictionRateLimit, evictionRateBurst)
-	factory, err := eviction.NewPodsEvictionRestrictionFactory(kubeClient, minReplicasForEvicition, evictionToleranceFraction)
+	lastInPlaceUpdateAttemptTimeMap := make(map[string]time.Time)
+	// TODO(maxcao13): Add inplace rate limits or just use eviction rate limits?
```
Thanks, I'll keep the TODO there, remove my name, and make the question a statement.

raywainman commented on vertical-pod-autoscaler/pkg/updater/restriction/pods_restriction_factory.go:
```diff
@@ -392,3 +325,70 @@ func setUpInformer(kubeClient kube_client.Interface, kind controllerKind) (cache
+}
+
+// CanEvictInPlacingPod checks if the pod can be evicted while it is currently in the middle of an in-place update.
+func CanEvictInPlacingPod(pod *apiv1.Pod, singleGroupStats singleGroupStats, lastInPlaceAttemptTimeMap map[string]time.Time, clock clock.Clock) bool {
```
Any chance we can move this into the pods_inplace_restriction.go file? That way it is closer to where it is used and this factory can be kept (mostly) agnostic of in-place stuff.

maxcao13 replied:
It's used in both restriction files for what it's worth, do you still want me to do that?

raywainman commented:
Overall LGTM. I would LOVE a second pair of eyes on this just given there is a lot happening here. @adrianmoisey, @omerap12, @voelzmo any chance I could get a second look?

Thanks a lot Max for patiently dealing with my waves of comments :)

omerap12 commented on vertical-pod-autoscaler/pkg/updater/logic/updater.go:
```diff
@@ -198,32 +213,88 @@ func (u *updater) RunOnce(ctx context.Context) {
+			err = u.inPlaceRateLimiter.Wait(ctx)
+			if err != nil {
+				klog.V(0).InfoS("In-place rate limiter wait failed for in-place resize", "error", err)
+				return
```
Here it makes sense to return because of the wait limit right?

maxcao13 replied:
I'm assuming so yes, that's just how it worked before for the eviction limiter.

omerap12 commented on vertical-pod-autoscaler/pkg/updater/restriction/pods_inplace_restriction.go:
```diff
@@ -0,0 +1,176 @@
+
+// TODO: Make these configurable by flags
+const (
+	// DeferredResizeUpdateTimeout defines the duration during which an in-place resize request
+	// is considered deferred. If the resize is not completed within this time, it falls back to eviction.
+	DeferredResizeUpdateTimeout = 5 * time.Minute
+
+	// InProgressResizeUpdateTimeout defines the duration during which an in-place resize request
+	// is considered in progress. If the resize is not completed within this time, it falls back to eviction.
+	InProgressResizeUpdateTimeout = 1 * time.Hour
```
Can we document the reasoning behind the chosen timeouts? (e.g., why the timeout is set to 5 minutes)

omerap12 replied:
TODO is fine :)

maxcao13 commented on vertical-pod-autoscaler/pkg/updater/logic/updater.go:
```diff
@@ -198,32 +213,88 @@ func (u *updater) RunOnce(ctx context.Context) {
+				klog.V(0).InfoS("In-place rate limiter wait failed for in-place resize", "error", err)
+				return
+			}
+			err := inPlaceLimiter.InPlaceUpdate(pod, vpa, u.eventRecorder)
+			if err != nil {
+				klog.V(0).InfoS("In-place update failed", "error", err, "pod", klog.KObj(pod))
+				return
```
You're right, it should probably continue, thanks for catching that.

maxcao13 replied:
Hmm... actually taking a look at it, most of the errors that are returned from the `InPlaceUpdate` function look like internal errors that shouldn't happen unless there is something wrong with the kube API or internally with the VPA. i.e., it seems like if one were to fail, then all subsequent in-place updates or evictions would fail too.

So I think it should be fine to keep it this way? If you think otherwise, please let me know.

omerap12 replied:
I think it might be better to use `continue` here, even though the current errors seem to happen consistently. This way, the code will be more resilient against sporadic issues and future updates to the InPlaceUpdate function that could introduce new error cases.

And since you mentioned "most of the errors" (not all), this seems like another reason to use `continue`.

What do you think, @raywainman?

raywainman replied:
Yeah I also noticed this when I reviewed and I went and took a look at the potential errors from the `InPlaceUpdate` function and concluded that they were all good reasons to stop

*However*, thinking about this some more... Omer, I think you are right here.. It would be a shame to completely stop processing a single VPA if something goes wrong with one pod.

What do we think about logging a warning here instead and doing a `continue`?

maxcao13 replied:
Okay, makes sense to me. Thanks for the feedback guys ^^

maxcao13 commented:
Is there anything else that needs to be addressed here?
/unhold

omerap12 commented:
lgtm, thank you! Excellent work, Max :) 
The next PR is for the end-to-end tests, right? I can't recall if we already merged it :)

maxcao13 commented:
Nope, I'll put that PR up now thanks Omer!

omerap12 commented:
> Nope, I'll put that PR up now thanks Omer!
> 
> (and it should be the last one before the big PR to main)

Great, Thanks!

maxcao13 commented:
PR is here and depends on this being merged: https://github.com/kubernetes/autoscaler/pull/8072

raywainman commented:
I know there are still some outstanding TODOs here but in the interest of unblocking this I think this is OK to go now. We can iterate from here. Really great work @maxcao13, thanks again.

/lgtm

/approve
